test/unit/compose/app.spec.ts

```diff
@@ -1671,4 +1672,91 @@ describe('compose/app', () => {
 			expectNoStep('kill', steps);
 		});
 	});
+
+	describe('update lock state behavior', () => {
+		it('should infer a releaseLock step if there are locks to be released before settling target state', async () => {
+			const services = [
+				await createService({ serviceName: 'server' }),
+				await createService({ serviceName: 'client' }),
+			];
+			const current = createApp({
+				services,
+				networks: [DEFAULT_NETWORK],
+			});
+			const target = createApp({
+				services,
+				networks: [DEFAULT_NETWORK],
+				isTarget: true,
+			});
+
+			const steps = current.nextStepsForAppUpdate(
+				{
+					...defaultContext,
+					locksTaken: new Map([[1, ['server', 'client']]]),
```

Comment:
Related to my previous comment about using map for keeping track of locks. While I agree is probably the right data structure, this code is not intuitive when looking at it without having the context of what the values inside the array are. If instead this code was

```
locksTaken: new LockRegistry([{appId: 1, services: ['server', 'client']}])
```

You would immediately get a sense of what the code does. The underlying implementation may still be using Map, but the interface is more clear

Reply:
Makes sense, will change. I was relying on Copilot to autocomplete the locksTakenMaps in test cases so didn't have to deal with this confusion as much, but you're right

Reply:
Updated `LocksTakenMap` to have more user-friendly instantiation args [in this commit](https://github.com/balena-os/balena-supervisor/pull/2234/commits/b5d5dd7a9837a8a939cea5d99fddc56f2d4840ac) (and added some tests)